### ISSUE TITLE
Work at the scale of twitter please(I worked there) with these features

### ISSUE THREAD

**Issue by deanorderly** [gradle/gradle#12802]
I break this feature request into the 'ultimate feature' gradle probably won't do, BUT then features that gradle could support to make the ultimate feature easier to obtain for us mono-repo people.

Gradle seems to offer no option to run a gradlew monobuild which would
1. detect changed files (gradle probably will never do this?)
2. map changed files to projects(gradle could do this...that would be so nice)
3. list all projects affected (the changed projects, the projects that depend on those transitively or not) 
4. force reverse declarations of dependencies for critical mono-repo speed.  Currently you must scan the ENTIRE mono-repo to figure out the web and this takes time for repos the size of twitter.
5. (there is additional info below on this one)

Let's talk #1.  This is a git call to determine files change (typically since master was forked since most people submit remote branches that CI merges into master IF things pass)

#2.  So, I can easily write a get changed files on git(and someone else can do SVN, etc).  After that, I have no way to give gradle a file to tell me the projects that have changed OR better yet, the projects that have changed AND all the dependencies

#3.  If I have a list of projects, is there a way to list ALL projects affected in a way that tells me WHICH projects i can build in parallel and which ones I have to build serially.  In mono-repo, this is huge for very large changes that have to kick off many CI build.

#4 next, to be VERY fast in a mono-repo, you have to know 'who depends on changed projects' transitively or not .  In gradle composite builds there is no setting for 'forceBiDirectionalDependencyDeclaration'.  If you had this, clients of my library would have to not only depend on my library BUT would have to change my libraries settings.xml(or some file) to say this libary C depends on my library.  With this, you get a HUGE added benefit.  It requires other teams that use my library to go through code review on my team so 1. we know we have a new customer and 2. we can reject them and refer them to a new library if needed.  Currently at twitter, this wasn't possible so you can imagine, people can just depend on anything they want.

#5.  In a mono-repo, library developers will typically make a change and use CI to build ALL projects as they affect way too many changes in the company.  HOWEVER, there are many developers in the company who want to run ./gradlew monobuild and for it to build any and all affected projects locally as they know it will not take that long.  If they are mistaken, welll, at twitter, we would just kill it and then submit it to the CI builds to run them all in parallel and such.

Currently, one has to write a TON of code to get a mono-build running and gradle should really handle that better.  I know one open source project we stole some bash scripting from outputs the build order of all projects needing to be built like so

{project1} {project2} {project3}
{project4}
{project5}

where the first line of 3 projects can be built in parallel BUT project 4 depend on one of those projects.  CLEARLY, this is not the best solution as if project1 finishes and project 4 only depends on that one project, it still waits for all 3.  Ideally, we could ask gradle the info perhaps at each step....not sure.

Anyways, help on making it easier to create mono-repos would be GREAT!!!!!

**Comment by ddimtirov** [comment]
What is the difference between this and `gradle -t` ?

**Comment by deanorderly** [comment]
@ddimtirov I must be missing something.  We need CI to start and kick off gradle build to build what has changed from the last commit on master.  Locally, we want to do whatever CI does to make sure it will pass CI before pushing to CI.

Also, I ran xgradlew -t and it didn't do anything.  I am not sure which directory I would run it from.

We have this structure

- monorepo
    - java
       -servers
          -server1
          -server2
          - ...
          -server30
      - libraries
          -library1
          -library2
          -etc

Building the entire thing would take quite some time so we want to run only what has changed and libraries and servers that depend on those changes.  I ran xgradlew -t in monorepo/java and it did nothing.  I ran it in server1 and it output "Exiting continuous build as no executed tasks declared file system inputs."

BUT I suspect running it on server1 would be very bad as it would only be compiling server1 and everything 'I depend on".  If I then change library1, there are about 5 servers that depend on library1 that should be built to make sure I didn't break them.  That of course includes server1 which I am currently working on.

**Comment by deanorderly** [comment]
oh, it works when I do xgradlew -t build BUT only for that one project!!!!  That's the thing, it needs to build ALL projects that CHANGED, not just server1 and what I depend on.  It also needs to build all projects that DEPEND on those projects.  It's a reverse tree from what build systems are used to which bazel and pants do nicely.

**Comment by ddimtirov** [comment]
It works for the dependencies of the tasks you are running (source sets and others). Kick a full monorepo build with -t and you would get what you described.

**Comment by ov7a** [comment]
Sorry for the late reply.

This issue will be closed as a duplicate of

* https://github.com/gradle/gradle/issues/27501

Please add your use case and 👍 to that issue.

If you think our analysis is wrong, please provide us with more detailed information explaining why.

***

Gradle mostly has features to support the described scenario: file system watching, continuous build, configuration cache, and build cache. The interesting part is covered already by #27501.
